crates/edr_solidity_tests/src/runner.rs

```diff
@@ -95,42 +106,211 @@ impl<'a> ContractRunner<'a> {
+        let test_results = functions
+            .par_iter()
+            .map(|&func| {
+                let _guard = handle.enter();
+
+                let sig = func.signature();
+
+                let setup = setup.clone();
+                let should_fail = self.test_fail && func.is_test_fail();
+                let res = if func.is_invariant_test() {
+                    let runner = test_options.invariant_runner();
+                    let invariant_config = test_options.invariant_config();
+                    self.run_invariant_test(
+                        executor.clone(),
+                        runner,
+                        setup,
+                        invariant_config.clone(),
+                        func,
+                        self.known_contracts,
+                        identified_contracts.as_ref().unwrap(),
+                    )
+                } else if func.is_fuzz_test() {
+                    debug_assert!(func.is_test());
+                    let runner = test_options.fuzz_runner();
+                    let fuzz_config = test_options.fuzz_config();
+                    self.run_fuzz_test(
+                        executor.clone(),
+                        func,
+                        should_fail,
+                        runner,
+                        setup,
+                        fuzz_config.clone(),
+                    )
+                } else {
+                    debug_assert!(func.is_test());
+                    self.run_test(executor.clone(), func, should_fail, setup)
+                };
+
+                (sig, res)
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        let duration = start.elapsed();
+        let suite_result = SuiteResult::new(
+            duration,
+            test_results,
+            warnings,
+            self.contract.libraries.clone(),
+        );
+        info!(
+            duration=?suite_result.duration,
+            "done. {}/{} successful",
+            suite_result.passed(),
+            suite_result.test_results.len()
+        );
+        suite_result
+    }
+
     /// Deploys the test contract inside the runner from the sending account,
     /// and optionally runs the `setUp` function on the test contract.
-    pub fn setup(&mut self, setup: bool) -> TestSetup {
-        match self._setup(setup) {
-            Ok(setup) => setup,
-            Err(err) => TestSetup::failed(err.to_string()),
-        }
+    fn setup(&self, executor: &mut Executor, needs_setup: bool) -> TestSetup {
+        self._setup(executor, needs_setup)
+            .unwrap_or_else(|err| TestSetup::failed(err.to_string()))
     }
 
-    fn _setup(&mut self, setup: bool) -> Result<TestSetup> {
-        trace!(?setup, "Setting test contract");
+    fn _setup(&self, executor: &mut Executor, needs_setup: bool) -> Result<TestSetup> {
+        trace!(?needs_setup, "Setting test contract");
```

Comment:
Question: What does `?` mean here?

Reply:
It's short hand for recording fields for structured tracing. Docs here (need to scroll down a bit): https://docs.rs/tracing/latest/tracing/#recording-fields

Reply:
Thanks!